### ISSUE TITLE
Map high-ASCII chars in DOS filenames to UTF-8 on the host

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#2619]
FAT16 allows all "high ASCII" characters (128-255) to be used in 8.3 filenames under MS-DOS ([reference](https://en.wikipedia.org/wiki/8.3_filename#Directory_table)).

That's all the funky symbols in the lower half of this chart showing the standard default CP437 charset:

<img width="484" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/44b3275b-a740-4f70-a045-2de4a73a64cd">

But who would be mad enough to do such a thing? All the cool demoscene kids from the 90s, brother! 😎  They thought that's a s000per kewl l337 h4xxx0r thing to do! 🤘🏻 

**cUz dAt $h!T iZ Th3rE 2 B UzEd, br0!!!!!!!111**

Case in point, I ran into such things in authentically preserved, mint condition ZIP files delivered to me straight from the 90s! As you can see, I was able to browse the contents of the ZIP file using DOS Navigators built-in archive viewer, but the unpacking failed because DOSBox was not able to map the funky characters to host-side filenames:

<img width="971" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/24b08c77-7141-46e8-a208-68188cb9dc96">

There's some complaints about the failed conversion in the logs:

<img width="710" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/040a03d5-c4a7-46b0-a9b5-77fb5fdef1d7">

Another example:

<img width="923" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/47fda826-4df8-457e-9c64-f81ded0f0c05">

<img width="746" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/45f571ad-8cca-4670-bc2c-983c2e63127f">

Both ZIP files are attached at the end of the ticket.

Just sharing this comical value: the built-in macOS unpack feature in Finder failed to unpack `IMPHOB6.ZIP` with an "invalid format" error. I had better luck with `MYSTIC~1.ZIP`, but I ended up with some Asian lookin characters. Wut? Clearly, there aren't many ex-demosceners at Apple... Americans!

<img width="499" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/34189b6f-9ad5-4d70-97fa-0d965145fca9">

---

The good news is that all CP437 characters have UTF-8 equivalents, as [shown here](https://en.wikipedia.org/wiki/Code_page_437).

NTFS and Linux/Mac filesystems can handle UTF-8 in filenames just fine, so it's just a matter of simply mapping these special chars to UTF-8 and Bob's your uncle!

Seeing is believing:

<img width="1021" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/d80a6109-cbd1-4e11-93c8-c794e2089582">

Of course, we need bi-directional mapping of these filenames between DOSBox and the host filesystem.

PS: I'm sure someone will point out that we need to do this properly and support such mappings for every DOS codepage mankind has ever invented. Sure, knock yourself out and add full localisation it you care about such stuff (I don't). But let's start with CP437, even hardcode the CP437->UTF-8 mappings in the first iteration because that will take care of the 99.99% of the actual practical use cases.

---

ZIP files containing files with high ASCII chars in their name:

[IMPHOB6.ZIP](https://github.com/dosbox-staging/dosbox-staging/files/11787344/IMPHOB6.ZIP)
[MYSTIC~1.ZIP](https://github.com/dosbox-staging/dosbox-staging/files/11787345/MYSTIC.1.ZIP)

**Comment by rderooy** [comment]
indeed, especially in certain "scenes" the use of such characters was not uncommon. For what its worth, DOSBox-X seems to handle it just fine.

**Comment by FeralChild64** [comment]
@johnnovak 

> just a matter of simply mapping these special chars to UTF-8 and Bob's your uncle!

```c
#include "string_utils.h"

dos_to_utf8(input_string, output_string, 437);
```

> Of course, we need bi-directional mapping of these special chars between DOSBox and the host filesystem.

```c
utf8_to_dos(input_string, output_string, UnicodeFallback::Null, 437);
```

But there is a catch here: not all UTF-8 characters can be directly mapped to given DOS code page. The `UnicodeFallback::Null` will map such characters to 0, and it is up to the filesystem emulation code to deal with it.
I was thinking about replacing all the non-matching Unicode glyphs with underscores and generating short file name, for example:
`AÐ-BORÐA.TXT` (`TO-EAT.TXT` in Icelandic) -> `A_-BOR_A.TXT` -> `A_-BOR~1.TXT`

> I'm sure someone will point out that we need to do this properly and support such mappings for every DOS codepage mankind has ever invented. Sure, knock yourself out and add full localisation it you care about such stuff (I don't)

Definitions for over a 100 code pages are ready :D Unfortunately, some elements are still missing:
- we need to teach our emulated userspace to use our Unicode engine uppercase/lowercase mapping
- we need to teach commands like DIR and TREE how to sort file names containing national characters (our Unicode engine does not have a mechanism for this yet)
- we need to export case table and collation table to compliant DOS software - see INT 21h, function 0x65

**Comment by Grounded0** [comment]
You could use CP850's eight bit for this wackyness too. How can we detect which codepage was the files originally authored on and decoding the intended character? In school we were ordered to only use ASCII filenames and only idiots were seen as using Swedish or Finnish characters in filenames like ä, ö or å.

**Comment by FeralChild64** [comment]
We can't detect, it's the user that has to select appropriate code page. But once the file is on host OS filesystem, with UTF-8 characters in file names, problem is basically solved (we just need to make sure our "userspace" is compatible).

DOS was clearly prepared (or was being prepared) for national characters in file names, it's the software that didn't catch up. Nevertheless - full NLS in file names is still a long way to go, and this can always be an optional feature.

**Comment by Grounded0** [comment]
I realised we could just use host codepage and use that as an assumption. But that assumption may be wrong. Duh.

**Comment by rderooy** [comment]
> I realised we could just use host codepage as an assumption. But that assumption may be wrong.

Any modern OS will be using Unicode, so host codepage will tell you nothing. But you could probably try to guess what a suitable codepage would be based on host country/language/keyboard settings.

**Comment by Grounded0** [comment]
I meant codepage DOSBox was set for but yeah. If DOSBox is set for CP437 use CP437 or if DOSBox is set for CP850 use CP850.

**Comment by FeralChild64** [comment]
This is actually easy, just omit the 437 parameter and it will use the currently set code page. But, as I have already mentioned, our commands are not ready yet, even with CP 437 users might see glitches if they try to use characters 0x80-0xA5.

**Comment by johnnovak** [comment]
Guyz! I only care about them warez filenamez, like I said 😎 Very few sane people used non 7-bit ASCII chars in filenames. So the 99% use case is: if you're a "scene enthusiast", use CP437 explicitly (or don't as that's the default). Then according to that nice table on Wikipedia, all CP437 special chars have their UTF-8 mappings. Then if someone tries to retrieve files from ZIP files stupidly named using non-English alphabet letters or something, thst's the 1% use I guess; they'll need to use whatever code page they need and the bidirectional mapping will work for them as long as they keep using that same codepage. But that's a pretty rare use case I'd think. I was also taught at schooll to forget about accented chars in filenames...

But there is a catch here: not all UTF-8 glyphs can be directly mapped to given DOS code page. The UnicodeFallback::Null will map such characters to 0, and it is up to the filesystem emulation code to deal with it.
I was thinking about replacing all the non-matching Unicode glyphs with underscores and generating short file name, for example:
AÐ-BORÐA.TXT (TO-EAT.TXT in Icelandic) -> A_-BOR_A.TXT -> A_-BOR~1.TXT

That' as good as it gets. But we'd get no underscores for CP437, correct?

**Comment by johnnovak** [comment]
Unfortunately, some elements are still missing:

    we need to teach our emulated userspace to use our Unicode engine uppercase/lowercase mapping
    we need to teach commands like DIR and TREE how to sort file names containing national characters (our Unicode engine does not have a mechanism for this yet)
    we need to export case table and collation table to compliant DOS software - see INT 21h, function 0x65

Hmmm, so DOS Navigator seems to display high ASCII when browsing ZIP files just fine, incl. sorting. How is that? I guess it doesn't have to use INT21h for that at all then?

I'm wondering if /storing/ DOS filenames internally as UTF-8 is a good decision? I mean, it's a bit pointless as DOS can only control the /interpretation/ of 8-bit values, so essentially we could store only the raw 8-bit streams in C strings or even better, as a custom object that stores the byte-identical DOS filenames and has a "toUtf8" method the does the conversio in the current code page as needed (and yes, it would cache the results if multiple conversion are requested for the same codepage).

Would that not vastly simplify this whole sorting thing? Storing filenames as converted UTF-8 string, then trying to come up with an UTF-8 sorting scheme that accurately matches the DOS behaviour seems hugely backwards to me. We could simply use the fully documented raw 8-bit sorting rules that DOS uses?

Correct me if I'm wrong, but I don't see why UTF-8 /storage/ on the emulation side brings us, only complications. Only one codepage can be present at any given time, so this becomes a display and interpretation issue only, hence I think my wrapper solution with lazy UTF-8 conversion would be simpler (plus it would have a "fromUtf8" method to decode to 8-bit raw DOS side string using a given DOS codepage).

**Comment by FeralChild64** [comment]
DOS Navigator can probably handle graphic characters, but try this: in your host file explorer create directories `windows` and `éditeur` (editor in French), try to display their names with `dir /on` under DOSBox, code page 437. I bet that:
- it will display `éDITEUR` instead of `ÉDITEUR`
- it will sort directory names wrongly and display `WINDOWS` before `éDITEUR`

The problem with MS-DOS sorting rules is that they cover some 15 code pages only (we can display about 90), limited amount of countries (not even mentioning that since then some countries disappeared and some nations gained independence), and national sorting rules do change over time, too. I have looked into FreeDOS source code (https://github.com/FDOS/country/blob/master/country.asm), but their collation rules are also far from complete.

I don't want to pass UTF-8 file names to DOS, this has no chance for this to work. I want to extend the Unicode engine to:
- contain generic Unicode sorting rules
- contain country-tailored Unicode sorting rules
- be able to export these rules (at least most of them) to DOS collation table, covering specific country and specific code page
- be able to compare strings in any supported DOS encoding - as some sorting rules (like handling German ß or Czech/Slovak accented letters and digraphs) cannot be exported to DOS collation tables, they are not flexible enough - see https://en.wikipedia.org/wiki/Alphabetical_order

----

Don't get me wrong - if you just want some really basic UTF-8 file names support, without all this quirks handled properly, and it already improves some use case, go for it. A complete support is far too complicated to be done in a single PR.

**Comment by johnnovak** [comment]
> DOS Navigator can probably handle graphic characters, but try this: in your host file explorer create directories `windows` and `éditeur` (editor in French), try to display their names with `dir /on` under DOSBox, code page 437. I bet that:
> 
>     * it will display `éDITEUR` instead of `ÉDITEUR`
>     * it will sort directory names wrongly and display `WINDOWS` before `éDITEUR`

Yup, I get that.
 
> The problem with MS-DOS sorting rules is that they cover some 15 code pages only (we can display about 90), limited amount of countries (not even mentioning that since then some countries disappeared and some nations gained independence), and national sorting rules do change over time, too. I have looked into FreeDOS source code (https://github.com/FDOS/country/blob/master/country.asm), but their collation rules are also far from complete.

Technically true, but who cares? 😎 DOSBox is our "time machine" into the 80s/90s; I wouldn't expect it to have "knowledge" about what happened after about 2000. I want to get the same behaviour as I got in the 80/90s, job done 😄 

Maybe you care about political implications because you mentioned national independence, etc. Personally, I don't care about politics much and don't wanna mix that into the project, so if some country was called "X" in the 90s, I don't see the point renaming it to "Y", its current name. Same way I really don't want books to be rewritten or "corrected" because the current political climate has changed, etc.

Anyway, my view, and I view it as a time machine 😄 

> Don't get me wrong - if you just want some really basic UTF-8 file names support, without all this quirks handled properly, and it already improves some use case, go for it. A complete support is far too complicated to be done in a single PR - and I don't want anybody to think that this is just a matter of encoding conversion.

Yeah, I think you're overcomplicating it, and I doubt speakers of those 90 - 15 = 75 languages care, but if you can make it work, who am I to no 😄 Ultimately, I want the CP437 full charset work in filenames and my desires stop there. I get it that some people would also like to use other European codepages, Russian, Turkish, etc., but beyond that I think you're heavily getting into the extreme niche category... But maybe if your solution is generic enough, it doesn't matter. Anyway, good luck 😄

**Comment by kcgen** [comment]
There's a subset of DOS games that bail out during install because the installer tries to write "illegal" filenames.  Turns out they are illegal in DOSBox, but permitted on real DOS.  So eXo's team runs the install under a real DOS environment, then carries the files out to DOSBox where it's usually OK. Unfortunately I can't remember them.. :man_facepalming: , but this might solve those too.

**Comment by johnnovak** [comment]
> There's a subset of DOS games that bail out during install because the installer tries to write "illegal" filenames. Turns out they are illegal in DOSBox, but permitted on real DOS. So eXo's team runs the install under a real DOS environment, then carries the files out to DOSBox where it's usually OK. Unfortunately I can't remember them.. 🤦‍♂️ , but this might solve those too.

Yeah, and this is one of the reasons why I always use WinUAE with HDD image files. Filenames on the Amiga's file system can contain characters that are illegal in NTFS. You rarely run into them, but when you do, it's an insurmountable issue.

**Comment by FeralChild64** [comment]
> Technically true, but who cares?

Me :) Seriously, code pages like 852 causes a lot of issues with software drawing tables, for many languages much better ones were created. And these new FreeDOS code pages are not always for newly supported languages, for example CP667 is a Polish standard Mazovia from 80s (before Microsoft introduced code pages in MS-DOS), which was supported by many dot-matrix printers and thus survived in some places till early XXIth century, CP867 is a similar Kamenický standard for Czech and Slovak languages, also from the 80s. CP668 is a stripped-down version of CP852, still suitable for Polish language, but with better table drawing compatibility, etc.

Regarding the countries - if you ask me, I would like to support both historic ones, and current ones; I want both Slovenia and Yugoslavia possible to select and supported :)

**Comment by kcgen** [comment]
> CP667 is a Polish standard Mazovia from 80s (before Microsoft introduced code pages in MS-DOS), which was supported by many dot-matrix printers and thus survived in some places till early XXIth century, CP867 is a similar Kamenický standard for Czech and Slovak languages, also from the 80s. CP668 is a stripped-down version of CP852, still suitable for Polish language, but with better table drawing compatibility

I didn't realize that FreeDOS code-pages can (in this case) cause an anachronism compared to using these these period-correct highly tailored Polish code-pages. 

Thanks for sharing this interesting history, @FeralChild64!

**Comment by johnnovak** [comment]
Okay, you're the expert @FeralChild64 😄 I remember using some other codepage from time to time when I wanted to print Hungarian accented letters on my dot matrix printer under DOS, but that's the extent of my experience with localisation and non-English codepages in pre-Windows times 😄 

I remember people sometime giving me files with Hungarian accented chars in the filenames on floppies around 2000, so pre-UTF8 times and I quite hated them for it because that caused complications on my English Windows.

Anyway, looking forward to your maximalist solution 😎

**Comment by Torinde** [comment]
> > my way of handling this would be it: don't create filenames with non-ASCII chars on the host filesystem, period
> 
> Not a problem for the first iteration: just check for 0 in the output string of `dos_to_utf8`; if it's there, file name is not supported and we can ignore the file altogether. At least for now :)

For first iteration OK, but I hope you afterwards go for your initial idea "replacing all the non-matching Unicode glyphs with underscores" - I feel ignoring whole files is not good, it's better to have the file present even with a wrong filename.

Host -> DOSbox
What about romanization-like mechanism, but for symbols, e.g. table(s) with the "closest looking / best replacement" character in each of the supported codepages?

Here I think we should not assume Host and DOS have the same language/country/etc. - for various reasons people may use "US/English/Microsoft-default" for some of the settings on either side, so I think it's good if DOSbox can take into account as many of those parameters as possible (and I think Staging already uses the chosen Windows regional format to automatically configure DOS KEYB?) 

DOSbox -> Host
Since all DOS codepage are fully representable in Unicode here will be no problem with finding symbols. But not every symbol is supported by every host OS/filesystem. Currently we have already the "Printable 7-bit ASCII" equivalence table, but even that may be too permissive? The target set of characters should be limited to those supported for filenames/folder names in all of filesystems for macOS, Linux, Windows and also in SMB network shares and Joliet for CD/DVD/BD. If one of those targets forbids a character - we don't use it. 

SMB/Joliet bring another restriction - length of names and also paths (e.g. folder/subfolder/filename) - some implementations length restriction counts 2-byte characters as 2, thus reducing even further the length (e.g. I've seen cases with ~150 full path length restriction where due to 2-byte characters, the maximum is just half of that), so depending on the use-case - sometimes it's better to use "restricted printable ASCII" even when a proper match is available in Unicode 2-byte.


On the topic of period authentic codepages - I agree with @FeralChild64, "new" FreeDOS codepages are in some (most?) cases capturing the _actually used_ ones from the 80s/90s period. Cyrillization for me was NEVER handled by the Microsoft OS tools - not via KEYB, etc. - instead various 3rd party programs were used. Bringing Win3.x, Win9x, their DOS prompts (which behave differently from regular DOS sometimes), copy-pasting to/from each of those, printing from each of those, filenames, keyboard layouts, etc. - headache! Only from Win2K/XP onwards it became mostly a non-topic (except for keyboard layout, which was solved by Microsoft even later).

**Comment by FeralChild64** [comment]
@Torinde Actually, I would like to present them - with file name indicating that something is wrong. I have some doubts about replacing the characters in file names with something looking similarly, this looks dangerous to me - maybe as an option. Anyway - this won't happen anytime soon, I am currently busy with EDIT.COM replacement.

**Comment by kcgen** [comment]
I suppose there are two factors in play:

**A hard requirement** that a filename consisting of the high-ASCII characters (regardless of locale at the time of write), can be read back in the future (regardless of locale at the time of read). 

The filename on the hosts' filesystem could appear as utter garbage and random symbols.  That's file to meet this need on the DOS side.

**A soft requirement** is that the filename, when listed inside DOSBox Staging, is that it (ideally) could look "correct" given the users locale.  However, this cannot conflict with the prior hard requirement. 

---


All supported operating systems (Linux, macOS, and Window, and even BSD and others like Solaris) support unicode characters in the host filesystem.

I think this should be a hard assumption for this feature and we use it and expect it (with the usually "If file creation failed, then LOG_ERR(...)"), and we shouldn't have complex compromises for FAT16, and other outdated filesystems (similar to the reason we're not jumping through hoops to support Windows XP).

**Comment by johnnovak** [comment]
> I think we should use and expect this (with the usual: "If file creation failed, then LOG_ERR(...)"), and we shouldn't have complex compromises and workarounds to support outdated host filesystems (like FAT16), for similar reason we're not jumping through hoops to support Windows XP.

Oh, absolutely. The good thing is we don't support WinXP. All common filesystem in use on today's systems are fit for purpose, then if someone uses some esoteric filesystem where this feature would break, well, that's their problem...

> Perhaps a "sidecar" metadata file would be needed to store these extra details about each file. However, this cannot conflict or endanger the prior hard requirement.

I don't think that's possible or even desirable with DOS? Depending on what codepage you select in DOS, or what 3rd party program you use that interprets the raw bytes in the FAT16 filenames, the interpretation of those bytes change. So it would not make sense I think to store the "last used" CP in that sidecar or something.

Anyway, I trust @FeralChild64 will come up with something good, I start losing track of what's in scope and how things will be handled...

But yeah, those strict requirements are the most important; the OS side filename can actually be garbage I don't care much actually.

**Comment by johnnovak** [comment]
Guyz got creative. Unpack that! 😎 

<img width="548" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/6c799173-687d-41ce-a067-10d2328c994c">

I could not find a way to get that file out of the ZIP under DOS! Fortunately, Finder could extract it, although the file has gained Russian citizenship in the process!

<img width="734" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/3f1ee36d-a529-4615-999c-4f666b7d77c4">

**Comment by Grounded0** [comment]
This is an eXoDOS issue:

- #3740

**Comment by Grounded0** [comment]
Dropping https://github.com/dosbox-staging/dosbox-staging/labels/game%20compatibility since the game (GATO) itself runs properly.

**Comment by johnnovak** [comment]
> Dropping [ game compatibility](https://github.com/dosbox-staging/dosbox-staging/labels/game%20compatibility) Issues related to a specific game since GATO (1984) itself runs properly even if the installer fails.

Yeah we don't need "game compatibility" for this one as this is purely about filename encodings. I have no idea where "GATO (1984)" was coming from...

**Comment by Grounded0** [comment]
> Yeah we don't need "game compatibility" for this one as this is purely about filename encodings. I have no idea where "GATO (1984)" was coming from...

- #3740

**Comment by johnnovak** [comment]
Updated implementation tips:

- https://github.com/dosbox-staging/dosbox-staging/pull/4117#issuecomment-2558258928

**Comment by Torinde** [comment]
I'm trying to test #4526, but so far experienced the following:
- `copy con ЯА` results in Staging closing without any dialog box - only the log has `EXCEPTION: Standard library exception: No mapping for the Unicode character exists in the target multi-byte code page.` (but unless you run from cmd.exe the [console window disappears too quickly](https://github.com/dosbox-staging/dosbox-staging/discussions/4032) to see the message)
- I managed to get in a situation where closing the console window results in a new instance of Staging being automatically fired up. If I close it, a new one appears, etc. And because I misclicked on the dosbox.exe too many times I had multiple of those firing up/closing in parallel.... and I can't stop/close them, almost like a maleware :) Eventually I tried using "exit" instead of closing-via-X (on the window bar) and that made them stop re-appearing
- With codepage 3021 in Staging - filenames on the host containing equivalents of 3021 high-ascii characters don't appear properly in Staging:
<img width="591" height="601" alt="Image" src="https://github.com/user-attachments/assets/4830b1f0-6a06-49f4-9ee1-15579f4bf58c" />

**Comment by johnnovak** [comment]
> I'm trying to test [#4526](https://github.com/dosbox-staging/dosbox-staging/pull/4526), but so far experienced the following:

You've kinda misunderstood the purpose of that PR. It's for handling UTF-8 chars from the host going _into_ Staging, "on the way" in. Nothing to do with special chars in the emulated filesystem.

**Comment by weirddan455** [comment]
> copy con ЯА results in Staging closing without any dialog box - only the log has EXCEPTION: Standard library exception: No mapping for the Unicode character exists in the target multi-byte code page.

This is fixed as of #4764

I'm planning to fix the other issues mentioned here after next release and I was just looking further into this now.

### Current situation on Linux (and probably Mac):

Extracting the DOS .zips linked on the OP on the host results in a UTF-8 encoded filename. If you copy the extracted files to your DOS drive, they will show up but have mangled names as DOSBox doesn't handle UTF-8 filenames on the host.

Extracting the DOS .zips using a DOS program like PKUNZIP from inside DOSBox Staging works as expected inside DOSBox. On the host side, the filenames show up like `''$'\376''SKYTOP'$'\376''.NOW'` which is invalid UTF-8.

### Current situation on Windows:

Extracting using PKUNZIP does convert the filename to UTF-16 and it will show up correctly on the host. However, any files using high-ASCII bytes will not show up at all in a DOSBox directory listing. There's some code in drive cache that's using ANSI Windows calls that needs to be cleaned up.

#4764 unintentionally changed behavior here as before that PR high-ASCII filenames would show up mangled and extracting with PKUNZIP threw an exception and crashed. Crash is fixed now but directory listing code needs a fix.

### Plan moving forward (after next release):

**Note: This will cause incompatibility with existing DOS drives containing high-ASCII filenames. They will need to be re-named on the host to UTF-8 (Linux/Mac) or UTF-16 (Windows).**

The drive cache contains a mapping between `origname` (filename on the host) and `shortname` (filename as shown in DOS). This is mostly to handle long file names on the host and map them to a short name that DOS can handle.

On Windows, `origname` needs to change from `char` to `wchar_t` and move from legacy ANSI WIN32 calls to the wide variant like we're already doing in `fs_utils_win32.cpp`.

Before generating `shortname`, convert it to codepage 437. Ignore or use substitution characters for anything that doesn't cleanly convert.

**Comment by johnnovak** [comment]
Looking forward to this @weirddan455 ! 🎉 

> **Note: This will cause incompatibility with existing DOS drives containing high-ASCII filenames. They will need to be re-named on the host to UTF-8 (Linux/Mac) or UTF-16 (Windows).**

That's fine. We can put a warning about that in the release notes, and the good news is, only a small handful of games would be affected by this (eXoDOS has a low single-digit number of games running from hard disk images to work around this). It's more prevalent in the demoscene with those old ZIP files from the 90s that can encode such funky chars.

**Comment by Torinde** [comment]
> Before generating `shortname`, convert it to codepage 437. Ignore or use substitution characters for anything that doesn't cleanly convert.

How will be handled cases were the high-ASCII chars are intended to be seen in non-437 codepage? Like filenames for languages with extra characters or non-Latin alphabet:
- host UTF8/16 filename -> currently active display codepage in DOSBox at the moment of filename reading/caching
- DOS codepage for filenames -> host UTF8/16 filename. DOS codepage for filenames probably can be assumed to be the same as the currently active display codepage in DOSBox at the moment of file creation/rename
- if someone does a lot of DOS display codepage switches, etc. - they can be considered "advanced user" who takes care himself of side effects (e.g. unzips Greek files while DOSBox is set for 437 - gets 437 characters on the host and in DOSBox, even if they switch to Greek codepage later. If they want Greek characters they should switch to Greek display before unzipping).

**Comment by johnnovak** [comment]
> How will be handled cases were the high-ASCII chars are intended to be seen in non-437 codepage? Like filenames for languages with extra characters or non-Latin alphabet:

You're not reading, man. You're not reading the tickets you're spamming with these infodumps... I guess you'll never learn 😐

Some help. From my top level ticket:

> But let's start with CP437, even hardcode the CP437->UTF-8 mappings in the first iteration because that will take care of the 99.99% of the actual practical use cases.

From @weirddan455 latest response:

> Before generating shortname, convert it to codepage 437. Ignore or use substitution characters for anything that doesn't cleanly convert.

Can you _not_ post noise to this issue ticket as per your standard practice? If you keep doing it, I'll need to hide your comments and/or mark them as spam when it's obviously clear you do not understand the agreed scope of the ticket. I really don't wanna play kindergarten cop here... It's super demotivating when some guy with way too much free time (that's you) posts all these weirdo niche issues/use cases _that you never intend to use anyway (!)_, burning up our valuable time and energy just parsing what you wrote and responding to it. Like I'm doing now.

Actually, I'll start hiding these comments of yours now. This gotta stop.

**Comment by Grounded0** [comment]
😁

Deploy the original CP437; once that is complete, one can address these granularities—provided any energy remains. In elementary school, we were specifically instructed to adhere to 7-bit ASCII to avoid the DOS 8-bit 'wormhole' (e.g., excluding Ä, Ö, and Å from filenames).

It is a pity so few operate with the Finnish engineering mindset of maintaining minimum required complexity.

**Comment by johnnovak** [comment]
> It is a pity so few operate with the Finnish engineering mindset of maintaining minimum required complexity.

Absolutely. Start small, get the 80-90% use case right, _then_ build upon it.

When in doubt, underengineer.

**Comment by johnnovak** [comment]
Did this go anywhere @weirddan455 ?

**Comment by weirddan455** [comment]
I'm working on it. Got delayed more than I would like due to life stuff.
